{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76fa8e67",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:51:19.917808Z",
     "iopub.status.busy": "2025-03-18T09:51:19.916921Z",
     "iopub.status.idle": "2025-03-18T09:51:19.959451Z",
     "shell.execute_reply": "2025-03-18T09:51:19.958718Z"
    },
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/home/lmf/anaconda3/envs/Mosaic_workflow/lib/python3.9/site-packages', '/home/lmf/.cache/snakemake/snakemake/source-cache/runtime-cache/tmpfoxbczqf/file/home/lmf/apps/MOSAIC/mosaic/rules/../notebooks', '/home/lmf/apps/MOSAIC/mosaic/rules/../notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95\\xba\\x16\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8cQ/home/lmf/DANIEL/VIRAL_METAGENOMES/05_vOTUs/checkV_merged_quality_summary.tot.txt\\x94\\x8cq/home/lmf/DANIEL/VIRAL_METAGENOMES/05_vOTUs/combined_positive_viral_contigs_derreplicated_rep_seq.tot_95-85.clstr\\x94\\x8ci/home/lmf/DANIEL/VIRAL_METAGENOMES/05_vOTUs/combined_positive_viral_contigs.tot_derreplicated_cluster.tsv\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x0emerged_summary\\x94K\\x00N\\x86\\x94\\x8c\\x0ccluster_file\\x94K\\x01N\\x86\\x94\\x8c\\x16derreplicated_clusters\\x94K\\x02N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh\\x18\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h\\x1e)}\\x94\\x8c\\x05_name\\x94h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bh\\x10h\\nh\\x12h\\x0bh\\x14h\\x0cub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8cL/home/lmf/DANIEL/VIRAL_METAGENOMES/05_vOTUs/vOTU_clustering_rep_list.tot.csv\\x94\\x8c`/home/lmf/DANIEL/VIRAL_METAGENOMES/05_vOTUs/vOTU_clustering_rep_list_checkv_per_category.tot.csv\\x94\\x8cK/home/lmf/DANIEL/VIRAL_METAGENOMES/05_vOTUs/new_references_clusters.tot.csv\\x94e}\\x94(h\\x0e}\\x94(\\x8c\\x0frepresentatives\\x94K\\x00N\\x86\\x94\\x8c\\x11checkv_categories\\x94K\\x01N\\x86\\x94\\x8c\\x0cnew_clusters\\x94K\\x02N\\x86\\x94uh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bh1h,h3h-h5h.ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94(]\\x94(\\x8c\\x06B_C1_V\\x94\\x8c\\x06B_C2_V\\x94\\x8c\\x06B_C3_V\\x94\\x8c\\x06B_C4_V\\x94\\x8c\\tB_CIT1B_V\\x94\\x8c\\tB_CIT1F_V\\x94\\x8c\\tB_CIT2B_V\\x94\\x8c\\tB_CIT2F_V\\x94\\x8c\\tB_CIT3B_V\\x94\\x8c\\tB_CIT3F_V\\x94\\x8c\\tB_CIT4B_V\\x94\\x8c\\tB_CIT4F_V\\x94\\x8c\\tB_CNT1B_V\\x94\\x8c\\tB_CNT1F_V\\x94\\x8c\\tB_CNT2B_V\\x94\\x8c\\tB_CNT2F_V\\x94\\x8c\\tB_CNT3B_V\\x94\\x8c\\tB_CNT3F_V\\x94\\x8c\\tB_CNT4B_V\\x94\\x8c\\tB_CNT4F_V\\x94\\x8c\\tB_OIT1B_V\\x94\\x8c\\tB_OIT1F_V\\x94\\x8c\\tB_OIT2B_V\\x94\\x8c\\tB_OIT2F_V\\x94\\x8c\\tB_OIT3B_V\\x94\\x8c\\tB_OIT3F_V\\x94\\x8c\\tB_OIT4B_V\\x94\\x8c\\tB_OIT4F_V\\x94\\x8c\\tB_ORT1B_V\\x94\\x8c\\tB_ORT1F_V\\x94\\x8c\\tB_ORT2B_V\\x94\\x8c\\tB_ORT2F_V\\x94\\x8c\\tB_ORT3B_V\\x94\\x8c\\tB_ORT3F_V\\x94\\x8c\\tB_ORT4B_V\\x94\\x8c\\tB_ORT4F_V\\x94\\x8c\\x07RS_C1_V\\x94\\x8c\\x07RS_C2_V\\x94\\x8c\\x07RS_C3_V\\x94\\x8c\\x07RS_C4_V\\x94\\x8c\\nRS_CIT1B_V\\x94\\x8c\\nRS_CIT1F_V\\x94\\x8c\\nRS_CIT2B_V\\x94\\x8c\\nRS_CIT2F_V\\x94\\x8c\\nRS_CIT3B_V\\x94\\x8c\\nRS_CIT3F_V\\x94\\x8c\\nRS_CIT4B_V\\x94\\x8c\\nRS_CIT4F_V\\x94\\x8c\\nRS_CNT1B_V\\x94\\x8c\\nRS_CNT1F_V\\x94\\x8c\\nRS_CNT2B_V\\x94\\x8c\\nRS_CNT2F_V\\x94\\x8c\\nRS_CNT3B_V\\x94\\x8c\\nRS_CNT3F_V\\x94\\x8c\\nRS_CNT4B_V\\x94\\x8c\\nRS_CNT4F_V\\x94\\x8c\\nRS_OIT1B_V\\x94\\x8c\\nRS_OIT1F_V\\x94\\x8c\\nRS_OIT2B_V\\x94\\x8c\\nRS_OIT2F_V\\x94\\x8c\\nRS_OIT3B_V\\x94\\x8c\\nRS_OIT3F_V\\x94\\x8c\\nRS_OIT4B_V\\x94\\x8c\\nRS_OIT4F_V\\x94\\x8c\\nRS_ORT1B_V\\x94\\x8c\\nRS_ORT1F_V\\x94\\x8c\\nRS_ORT2B_V\\x94\\x8c\\nRS_ORT2F_V\\x94\\x8c\\nRS_ORT3B_V\\x94\\x8c\\nRS_ORT3F_V\\x94\\x8c\\nRS_ORT4B_V\\x94\\x8c\\nRS_ORT4F_V\\x94e\\x8c-/home/lmf/DANIEL/VIRAL_METAGENOMES/03_CONTIGS\\x94\\x8c./home/lmf/DANIEL/VIRAL_METAGENOMES/04_VIRAL_ID\\x94\\x88\\x89e}\\x94(h\\x0e}\\x94(\\x8c\\x07samples\\x94K\\x00N\\x86\\x94\\x8c\\ncontig_dir\\x94K\\x01N\\x86\\x94\\x8c\\tviral_dir\\x94K\\x02N\\x86\\x94\\x8c\\x0bsubassembly\\x94K\\x03N\\x86\\x94\\x8c\\x0ecross_assembly\\x94K\\x04N\\x86\\x94uh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bh\\x91hDh\\x93h\\x8dh\\x95h\\x8eh\\x97\\x88h\\x99\\x89ub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94\\x8c\\x03tot\\x94a}\\x94(h\\x0e}\\x94\\x8c\\x08sampling\\x94K\\x00N\\x86\\x94sh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94b\\x8c\\x08sampling\\x94h\\xa8ub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x04/tmp\\x94e}\\x94(h\\x0e}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bh\\xbfK\\x01h\\xc1K\\x01h\\xc3h\\xbcub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8cM/home/lmf/DANIEL/VIRAL_METAGENOMES/NOTEBOOKS/05_vOTU_representative.tot.ipynb\\x94a}\\x94(h\\x0e}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bh\\xd5h\\xd2ub\\x8c\\x06config\\x94}\\x94(\\x8c\\tinput_dir\\x94\\x8c./home/lmf/DANIEL/VIRAL_METAGENOMES/00_RAW_DATA\\x94\\x8c\\tkraken_db\\x94\\x8c\\x1b/home/lmf/db/KRAKEN/kraken/\\x94\\x8c\\rkrakenUniq_db\\x94\\x8c\\x1e/home/lmf/db/KRAKEN/KrakenUniq\\x94\\x8c\\necc_memory\\x94M\\x80>\\x8c\\x0bsubassembly\\x94\\x88\\x8c\\x1cadditional_reference_contigs\\x94\\x8c7/home/lmf/DANIEL/VIRAL_METAGENOMES/WRB_75_contigs.fasta\\x94\\x8c\\x0eassembly_stats\\x94\\x88\\x8c\\x0bsubsampling\\x94\\x88\\x8c\\x0fmin_votu_length\\x94M\\x10\\'\\x8c\\x0crun_vcontact\\x94\\x88\\x8c\\x0bimgvr_blast\\x94\\x88\\x8c\\x08run_DRAM\\x94\\x88\\x8c\\x11microbial_spacers\\x94\\x8c\\\\/home/lmf/DANIEL/MG-Assembly/ALL_MG_megahit_2000bp_derreplicated_minced_predicted_spacers.fa\\x94\\x8c\\x0bresults_dir\\x94\\x8c\\x00\\x94\\x8c\\nmetagenome\\x94\\x88\\x8c\\tmicrobial\\x94\\x89\\x8c\\x08sourmash\\x94\\x89\\x8c\\x0bkey_samples\\x94\\x8c\\x14key_samples_list.csv\\x94\\x8c\\x10operating_system\\x94\\x8c\\x05linux\\x94\\x8c\\x0bforward_tag\\x94\\x8c\\x02R1\\x94\\x8c\\x0breverse_tag\\x94\\x8c\\x02R2\\x94\\x8c\\x0cnanopore_tag\\x94\\x8c\\x08nanopore\\x94\\x8c\\x0fnanopore_pooled\\x94\\x89\\x8c\\x14nanopore_pooled_name\\x94\\x8c\\x03raw\\x94\\x8c\\x0elong_assembler\\x94\\x8c\\x04canu\\x94\\x8c\\x0bgenome_size\\x94\\x8c\\x0320m\\x94\\x8c\\x13trimmomatic_leading\\x94\\x8c\\x0220\\x94\\x8c\\x14trimmomatic_trailing\\x94\\x8c\\x0220\\x94\\x8c\\x17trimmomatic_window_size\\x94\\x8c\\x014\\x94\\x8c\\x1atrimmomatic_window_quality\\x94\\x8c\\x0220\\x94\\x8c\\x12trimmomatic_minlen\\x94\\x8c\\x0250\\x94\\x8c\\x10nanopore_quality\\x94\\x8c\\x0210\\x94\\x8c\\x0fnanopore_minlen\\x94M\\xe8\\x03\\x8c\\x0fnanopore_maxlen\\x94J\\x80\\x96\\x98\\x00\\x8c\\tVirSorter\\x94\\x89\\x8c\\x0ecross_assembly\\x94\\x89\\x8c\\nlong_index\\x94\\x89\\x8c\\x16representative_contigs\\x94h\\xf5\\x8c\\x08sra_list\\x94h\\xf5\\x8c\\x0eextract_mapped\\x94\\x89\\x8c\\x07p_ident\\x94\\x8c\\x0295\\x94\\x8c\\x0bkmer_window\\x94\\x8c\\x0525000\\x94\\x8c\\x08min_norm\\x94\\x8c\\x010\\x94\\x8c\\x08max_norm\\x94\\x8c\\x03100\\x94\\x8c\\rmax_subsample\\x94\\x8c\\x0850000000\\x94\\x8c\\x07min_len\\x94\\x8c\\x041000\\x94\\x8c\\x07min_cov\\x94\\x8c\\x012\\x94\\x8c\\x11contaminants_list\\x94h\\xf5\\x8c\\radapters_file\\x94\\x8c\\x0badapters.fa\\x94\\x8c\\x11ambiguous_mapping\\x94\\x8c\\x03all\\x94\\x8c\\x0ebasecalled_dir\\x94\\x8c\\x10data/basecalled/\\x94\\x8c\\x13reference_databases\\x94\\x8c\\x1eIMGVR GenBankViral RefSeqViral\\x94\\x8c\\x08IMGVR_db\\x94\\x8c!/home/lmf/db/IMG_VR_2022-12-19_7/\\x94\\x8c\\x0eRefSeqViral_db\\x94\\x8c6/home/lmf/db/RefSeqViral/RefSeq_viral_2024_09_03.fasta\\x94\\x8c\\x0bpharokka_db\\x94N\\x8c\\x08blast_db\\x94\\x8c\\x1fdb/ncbi/NCBI_viral_proteins.faa\\x94\\x8c\\x0cvirSorter_db\\x94\\x8c\\x1bdb/VirSorter/virsorter-data\\x94\\x8c\\x0ckraken_db_nt\\x94\\x8c\\x17/home/lmf/db/KRAKEN/nt/\\x94\\x8c\\x0fkraken_db_human\\x94\\x8c\\x13db/kraken2_human_db\\x94\\x8c\\x0ckraken_tools\\x94\\x8c\\x11tools/KrakenTools\\x94\\x8c\\x0bbracken_dir\\x94\\x8c\\rtools/Bracken\\x94\\x8c\\nweesam_dir\\x94\\x8c\\x0ctools/weeSAM\\x94\\x8c\\tcheckv_db\\x94\\x8c\\x11db/checkv-db-v1.5\\x94\\x8c\\tcheckm_db\\x94\\x8c\\x0cdb/checkm-db\\x94\\x8c\\x08iphop_db\\x94\\x8c\\x0bdb/iphop_db\\x94\\x8c\\rtaxmyphage_db\\x94\\x8c\\x10db/taxmyphage_db\\x94\\x8c\\x0csourmash_sig\\x94\\x8c7/home/lmf/db/SOURMASH/matrix_09_42_sourmash_k31.sig.zip\\x94\\x8c\\x10sourmash_rocksdb\\x94\\x8c;/home/lmf/db/SOURMASH/matrix_09_42_sourmash_k31.sig.rocksdb\\x94\\x8c\\x0csourmash_tax\\x94\\x8c9/home/lmf/db/SOURMASH/taxonomy/taxonomy_matrix_strain.csv\\x94\\x8c\\x08canu_dir\\x94\\x8c\\x1etools/canu-2.0/Linux-amd64/bin\\x94\\x8c\\x12PhaGCN_newICTV_dir\\x94\\x8c\\x14tools/PhaGCN_newICTV\\x94\\x8c\\tquast_dir\\x94\\x8c\\x11tools/quast-5.0.2\\x94\\x8c\\rvirSorter_dir\\x94\\x8c\\x0ftools/VirSorter\\x94\\x8c\\x07WTP_dir\\x94\\x8c\\ttools/WTP\\x94\\x8c\\rvirFinder_dir\\x94\\x8c\\x0ftools/VirFinder\\x94\\x8c\\x0eclusterONE_dir\\x94\\x8c\\x10tools/clusterONE\\x94\\x8c\\x0cvcontact_dir\\x94\\x8c\\x0ftools/vcontact2\\x94\\x8c\\x1areference_genomes_vcontact\\x94\\x8c\\x04None\\x94\\x8c\\rtaxonomy_file\\x94\\x8c1db/vcontact2/formatting_taxonomy_affiliations.csv\\x94\\x8c\\ngenomad_db\\x94\\x8c\\rdb/genomad_db\\x94\\x8c\\x07DRAM_db\\x94\\x8c\\ndb/DRAM_db\\x94\\x8c\\tgtdbtk_db\\x94\\x8c\\ndb/GTDB-Tk\\x94\\x8c\\nmmseqs_dir\\x94\\x8c\\rtools/MMseqs2\\x94\\x8c\\x08viga_dir\\x94\\x8c\\ntools/viga\\x94\\x8c\\tpiler_dir\\x94\\x8c\\x11tools/pilercr1.06\\x94\\x8c\\x07trf_dir\\x94\\x8c\\ttools/TRF\\x94\\x8c\\x0bvibrant_dir\\x94\\x8c\\rtools/VIBRANT\\x94\\x8c\\x0eviridic_folder\\x94\\x8c\\x17tools/viridic_v1.0_r3.6\\x94\\x8c\\x0fsra_toolkit_dir\\x94\\x8c tools/sratoolkit.2.10.0-ubuntu64\\x94\\x8c\\x07ALE_dir\\x94\\x8c\\ttools/ALE\\x94\\x8c\\x08wish_dir\\x94\\x8c\\ntools/WIsH\\x94\\x8c\\nminced_dir\\x94\\x8c\\x0ctools/minced\\x94\\x8c\\x07dion_db\\x94\\x8c\\x0cdb/dionSetDB\\x94u\\x8c\\x04rule\\x94\\x8c\\x1aselect_vOTU_representative\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8c//home/lmf/apps/MOSAIC/mosaic/rules/../notebooks\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = True; import os; os.chdir(r'/home/lmf/apps/MOSAIC/mosaic');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "85993d3b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:51:19.963398Z",
     "iopub.status.busy": "2025-03-18T09:51:19.963202Z",
     "iopub.status.idle": "2025-03-18T09:51:20.172974Z",
     "shell.execute_reply": "2025-03-18T09:51:20.171666Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8ff1e2e9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:51:20.178020Z",
     "iopub.status.busy": "2025-03-18T09:51:20.177752Z",
     "iopub.status.idle": "2025-03-18T09:51:20.182205Z",
     "shell.execute_reply": "2025-03-18T09:51:20.181499Z"
    }
   },
   "outputs": [],
   "source": [
    "input_merged_summary=snakemake.input.merged_summary\n",
    "input_cluster_file=snakemake.input.cluster_file\n",
    "input_derreplicated_clusters=snakemake.input.derreplicated_clusters\n",
    "output_representatives=snakemake.output.representatives\n",
    "output_checkv_categories=snakemake.output.checkv_categories\n",
    "output_new_clusters=snakemake.output.new_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f3218f32",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:51:20.185790Z",
     "iopub.status.busy": "2025-03-18T09:51:20.185599Z",
     "iopub.status.idle": "2025-03-18T09:51:32.168325Z",
     "shell.execute_reply": "2025-03-18T09:51:32.167385Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id_vOTU</th>\n",
       "      <th>contig_length_vOTU</th>\n",
       "      <th>checkv_quality_vOTU</th>\n",
       "      <th>completeness_vOTU</th>\n",
       "      <th>provirus_vOTU</th>\n",
       "      <th>warnings_vOTU</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1282758</th>\n",
       "      <td>RS_ORT2B_V_80_NODE_18783_length_1000_cov_2.478307</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Not-determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>no viral genes detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161202</th>\n",
       "      <td>B_CIT2B_V_NODE_54892_length_1000_cov_3.166138</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Not-determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>no viral genes detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161203</th>\n",
       "      <td>B_CIT2B_V_NODE_54896_length_1000_cov_3.005291</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Not-determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>no viral genes detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161205</th>\n",
       "      <td>B_CIT2B_V_NODE_54920_length_1000_cov_2.602116</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Not-determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>no viral genes detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161206</th>\n",
       "      <td>B_CIT2B_V_NODE_54932_length_1000_cov_2.344974</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>Not-determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>no viral genes detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1245868</th>\n",
       "      <td>RS_OIT3B_V_NODE_1_length_121268_cov_12.552548</td>\n",
       "      <td>121268.0</td>\n",
       "      <td>Complete</td>\n",
       "      <td>100.00</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565035</th>\n",
       "      <td>B_CNT3F_V_70_NODE_1_length_130741_cov_12.464258</td>\n",
       "      <td>130741.0</td>\n",
       "      <td>Complete</td>\n",
       "      <td>100.00</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>693290</th>\n",
       "      <td>B_CNT4F_V_NODE_1_length_143009_cov_13.424794</td>\n",
       "      <td>143009.0</td>\n",
       "      <td>Medium-quality</td>\n",
       "      <td>80.05</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156585</th>\n",
       "      <td>B_CIT2B_V_NODE_1_length_172123_cov_16.595224</td>\n",
       "      <td>172123.0</td>\n",
       "      <td>Complete</td>\n",
       "      <td>100.00</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576301</th>\n",
       "      <td>B_CNT3F_V_80_NODE_1_length_215226_cov_11.905062</td>\n",
       "      <td>215226.0</td>\n",
       "      <td>High-quality</td>\n",
       "      <td>94.54</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>229625 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            contig_id_vOTU  \\\n",
       "1282758  RS_ORT2B_V_80_NODE_18783_length_1000_cov_2.478307   \n",
       "161202       B_CIT2B_V_NODE_54892_length_1000_cov_3.166138   \n",
       "161203       B_CIT2B_V_NODE_54896_length_1000_cov_3.005291   \n",
       "161205       B_CIT2B_V_NODE_54920_length_1000_cov_2.602116   \n",
       "161206       B_CIT2B_V_NODE_54932_length_1000_cov_2.344974   \n",
       "...                                                    ...   \n",
       "1245868      RS_OIT3B_V_NODE_1_length_121268_cov_12.552548   \n",
       "565035     B_CNT3F_V_70_NODE_1_length_130741_cov_12.464258   \n",
       "693290        B_CNT4F_V_NODE_1_length_143009_cov_13.424794   \n",
       "156585        B_CIT2B_V_NODE_1_length_172123_cov_16.595224   \n",
       "576301     B_CNT3F_V_80_NODE_1_length_215226_cov_11.905062   \n",
       "\n",
       "         contig_length_vOTU checkv_quality_vOTU  completeness_vOTU  \\\n",
       "1282758              1000.0      Not-determined                NaN   \n",
       "161202               1000.0      Not-determined                NaN   \n",
       "161203               1000.0      Not-determined                NaN   \n",
       "161205               1000.0      Not-determined                NaN   \n",
       "161206               1000.0      Not-determined                NaN   \n",
       "...                     ...                 ...                ...   \n",
       "1245868            121268.0            Complete             100.00   \n",
       "565035             130741.0            Complete             100.00   \n",
       "693290             143009.0      Medium-quality              80.05   \n",
       "156585             172123.0            Complete             100.00   \n",
       "576301             215226.0        High-quality              94.54   \n",
       "\n",
       "        provirus_vOTU            warnings_vOTU  \n",
       "1282758            No  no viral genes detected  \n",
       "161202             No  no viral genes detected  \n",
       "161203             No  no viral genes detected  \n",
       "161205             No  no viral genes detected  \n",
       "161206             No  no viral genes detected  \n",
       "...               ...                      ...  \n",
       "1245868            No                      NaN  \n",
       "565035             No                      NaN  \n",
       "693290             No                      NaN  \n",
       "156585             No                      NaN  \n",
       "576301             No                      NaN  \n",
       "\n",
       "[229625 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read and preprocess the clustered_df DataFrame\n",
    "clustered_df = pd.read_csv(input_cluster_file, sep=\"\\t\", names=[\"rep\", \"mem\"])\n",
    "clustered_df[\"mem\"] = clustered_df[\"mem\"].str.split(\",\")\n",
    "clustered_df = clustered_df.explode('mem').reset_index(drop=True)\n",
    "\n",
    "# Read and preprocess the checkv_df_vOTUs DataFrame\n",
    "checkv_df_vOTUs = pd.read_csv(input_merged_summary, sep=\"\\t\",\n",
    "                            names= [\"contig_id_vOTU\", \"contig_length_vOTU\", \"provirus_vOTU\", \"proviral_length_vOTU\",\n",
    "                           \"gene_count_vOTU\", \"viral_genes_vOTU\", \"host_genes_vOTU\", \"checkv_quality_vOTU\",\n",
    "                           \"miuvig_quality_vOTU\", \"completeness_vOTU\", \"completeness_method_vOTU\",\n",
    "                           \"contamination_vOTU\", \"kmer_freq_vOTU\", \"warnings_vOTU\"])\n",
    "checkv_df_vOTUs = checkv_df_vOTUs[[\"contig_id_vOTU\", \"contig_length_vOTU\", \"checkv_quality_vOTU\",\n",
    "                                   \"completeness_vOTU\", \"provirus_vOTU\", \"warnings_vOTU\"]]\n",
    "checkv_df_vOTUs[\"contig_length_vOTU\"] = checkv_df_vOTUs[\"contig_length_vOTU\"].astype(float)\n",
    "checkv_df_vOTUs[\"completeness_vOTU\"] = checkv_df_vOTUs[\"completeness_vOTU\"].astype(float)\n",
    "checkv_df_vOTUs=checkv_df_vOTUs.groupby(\"contig_id_vOTU\").first().reset_index()\n",
    "\n",
    "# Create copies of the checkv_df_vOTUs DataFrame\n",
    "checkv_df_rep = checkv_df_vOTUs.copy()\n",
    "checkv_df_rep.columns = [\"contig_id_rep\", \"contig_length_rep\", \"checkv_quality_rep\", \"completeness_rep\",\n",
    "                         \"provirus_rep\", \"warnings_rep\"]\n",
    "\n",
    "# Merge DataFrames and drop unnecessary columns\n",
    "merged_df = clustered_df.merge(checkv_df_rep, left_on=\"rep\", right_on=\"contig_id_rep\", how=\"left\") \\\n",
    "                        .merge(checkv_df_vOTUs, left_on=\"mem\", right_on=\"contig_id_vOTU\", how=\"left\")\n",
    "merged_df = merged_df.drop(['contig_id_vOTU', 'contig_id_rep'], axis=1)\n",
    "merged_df[['completeness_vOTU', 'contig_length_vOTU', 'completeness_rep']] = merged_df[\n",
    "    ['completeness_vOTU', 'contig_length_vOTU', 'completeness_rep']].fillna(0)\n",
    "merged_df[\"completeness_vOTU\"] = merged_df[\"completeness_vOTU\"].astype(float)\n",
    "merged_df[\"contig_length_vOTU\"] = merged_df[\"contig_length_vOTU\"].astype(float)\n",
    "\n",
    "# Choose the contig with the highest completeness and the lowest contig length\n",
    "merged_df_best = merged_df[merged_df.groupby(['rep'], sort=False)['completeness_vOTU'].transform(max) == merged_df['completeness_vOTU']]\n",
    "merged_df_best = merged_df_best[merged_df_best.groupby(['rep'], sort=False)['contig_length_vOTU'].transform(min) == merged_df_best['contig_length_vOTU']]\n",
    "merged_df_best = merged_df_best.groupby(['rep']).first().reset_index()\n",
    "\n",
    "# Select final singletons and grouped contigs based on checkv quality\n",
    "final_singletons = merged_df_best[merged_df_best[\"rep\"] == merged_df_best[\"mem\"]][\"rep\"].tolist()\n",
    "\n",
    "merged_df_diff = merged_df_best[merged_df_best[\"rep\"] != merged_df_best[\"mem\"]]\n",
    "merged_df_diff = merged_df_diff.replace({'Complete': 4, 'High-quality': 3, 'Medium-quality': 2,'Low-quality': 1, 'Not-determined': 1})\n",
    "\n",
    "# Choose representatives based on checkv quality comparison\n",
    "merged_df_greatherthan = merged_df_diff[merged_df_diff[\"checkv_quality_rep\"] > merged_df_diff[\"checkv_quality_vOTU\"]]\n",
    "gr1 = merged_df_greatherthan[\"rep\"].tolist()\n",
    "\n",
    "merged_df_smallerthan = merged_df_diff[merged_df_diff[\"checkv_quality_rep\"] < merged_df_diff[\"checkv_quality_vOTU\"]]\n",
    "gr2 = merged_df_smallerthan[\"mem\"].tolist()\n",
    "\n",
    "merged_df_equal1 = merged_df_diff[(merged_df_diff[\"checkv_quality_rep\"] == merged_df_diff[\"checkv_quality_vOTU\"]) & \n",
    "                                       (merged_df_diff[\"warnings_rep\"]!=\"contig >1.5x longer than expected genome length\")]\n",
    "\n",
    "gr3 = merged_df_equal1[\"rep\"].tolist()\n",
    "\n",
    "merged_df_equal2 = merged_df_diff[(merged_df_diff[\"checkv_quality_rep\"] == merged_df_diff[\"checkv_quality_vOTU\"]) & \n",
    "                                       (merged_df_diff[\"warnings_rep\"]==\"contig >1.5x longer than expected genome length\")]\n",
    "gr4 = merged_df_equal2[merged_df_equal2[\"warnings_vOTU\"]==\"contig >1.5x longer than expected genome length\"][\"rep\"].tolist()\n",
    "\n",
    "gr5 = merged_df_equal2[merged_df_equal2[\"warnings_vOTU\"]!=\"contig >1.5x longer than expected genome length\"][\"mem\"].tolist()\n",
    "\n",
    "# merged_df_greatherthan.sort_values(by=\"contig_length_vOTU\")[-30:]\n",
    "representatives = final_singletons + gr1 + gr2 + gr3 +  gr4 + gr5\n",
    "final_df_rep = checkv_df_vOTUs[checkv_df_vOTUs['contig_id_vOTU'].isin(representatives)]\n",
    "\n",
    "# Sort the final_df_rep DataFrame by contig_length_vOTU column\n",
    "final_df_rep.sort_values(by=\"contig_length_vOTU\")                                                            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5846bd6c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:51:32.173951Z",
     "iopub.status.busy": "2025-03-18T09:51:32.173707Z",
     "iopub.status.idle": "2025-03-18T09:51:32.214620Z",
     "shell.execute_reply": "2025-03-18T09:51:32.213924Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>checkv_quality_vOTU</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Complete</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>High-quality</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low-quality</th>\n",
       "      <td>101803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Medium-quality</th>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Not-determined</th>\n",
       "      <td>127400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0\n",
       "checkv_quality_vOTU        \n",
       "Complete                 33\n",
       "High-quality             28\n",
       "Low-quality          101803\n",
       "Medium-quality          361\n",
       "Not-determined       127400"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df_rep.groupby(\"checkv_quality_vOTU\").size().to_csv(output_checkv_categories, index=True, header=False)\n",
    "final_df_rep.groupby(\"checkv_quality_vOTU\").size().to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "263dd02f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:51:32.219089Z",
     "iopub.status.busy": "2025-03-18T09:51:32.218883Z",
     "iopub.status.idle": "2025-03-18T09:51:40.276442Z",
     "shell.execute_reply": "2025-03-18T09:51:40.275055Z"
    }
   },
   "outputs": [],
   "source": [
    "derreplicated_df=pd.read_csv(input_derreplicated_clusters, sep=\"\\t\", names=[\"rep_d\", \"mem_d\"])\n",
    "derreplicated_df=derreplicated_df.merge(clustered_df, left_on=\"rep_d\", right_on=\"mem\")\n",
    "derreplicated_df=derreplicated_df[[\"rep\", \"mem_d\"]]\n",
    "# Make a copy of the original DataFrame `derreplicated_df` to avoid modifying it directly\n",
    "derreplicated_df_fixed = derreplicated_df.copy()\n",
    "\n",
    "# Create a set of unique values from the \"rep\" column in `derreplicated_df` for quick lookup\n",
    "rep_set = set(derreplicated_df[\"rep\"])\n",
    "\n",
    "# # Loop over each new reference (in `contig_id_vOTU` column) from `final_df_rep`\n",
    "for new_reference in final_df_rep[\"contig_id_vOTU\"]:\n",
    "    \n",
    "    # Check if the new reference is not already in the existing set of \"rep\" values\n",
    "    if new_reference not in rep_set:\n",
    "        \n",
    "        # Find the old reference in `derreplicated_df` where `mem` column matches the `new_reference`\n",
    "        # Get the corresponding value from the \"rep\" column\n",
    "        old_reference = derreplicated_df.loc[derreplicated_df[\"mem_d\"] == new_reference, \"rep\"].iloc[0]\n",
    "        \n",
    "        # Update `derreplicated_df_fixed` so that all occurrences of `old_reference` in \"rep\" are replaced by `new_reference`\n",
    "        derreplicated_df_fixed.loc[derreplicated_df_fixed[\"rep\"] == old_reference, \"rep\"] = new_reference\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "89708148",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:51:40.281696Z",
     "iopub.status.busy": "2025-03-18T09:51:40.281467Z",
     "iopub.status.idle": "2025-03-18T09:53:16.329256Z",
     "shell.execute_reply": "2025-03-18T09:53:16.327768Z"
    }
   },
   "outputs": [],
   "source": [
    "for rep, group in derreplicated_df_fixed.groupby(\"rep\"):\n",
    "    # Check if `rep` does not start with \"P_\" or \"E_\"\n",
    "    if not rep.startswith(\"P_\") and not rep.startswith(\"E_\"):\n",
    "        \n",
    "        # Check if any value in the \"mem_d\" column starts with \"P_\" or \"E_\"\n",
    "        if group[\"mem_d\"].str.startswith(\"P_\").any() or group[\"mem_d\"].str.startswith(\"E_\").any():\n",
    "            \n",
    "            # Print the matching `rep` value\n",
    "            print(f\"Match found for rep: {rep}\")\n",
    "            \n",
    "            # Find values in \"mem_d\" that start with \"P_\" or \"E_\"\n",
    "            matching_mem_d = group[\"mem_d\"][group[\"mem_d\"].str.startswith(\"P_\") | group[\"mem_d\"].str.startswith(\"E_\")]\n",
    "            \n",
    "            # Collect `mem_d` and `contig_length_rep` pairs for sorting\n",
    "            mem_d_lengths = []\n",
    "            \n",
    "            # Loop over each matching mem_d value to get the corresponding contig length from checkv_df_rep\n",
    "            for mem_d in matching_mem_d:\n",
    "                # Find the contig length for each matching `mem_d` in `checkv_df_rep`\n",
    "                contig_length = checkv_df_rep.loc[checkv_df_rep[\"contig_id_rep\"] == mem_d, \"contig_length_rep\"]\n",
    "                if not contig_length.empty:\n",
    "                    mem_d_lengths.append((mem_d, contig_length.iloc[0]))\n",
    "\n",
    "            # Sort the list by `contig_length_rep` (second item in tuple)\n",
    "            mem_d_lengths.sort(key=lambda x: x[1])\n",
    "\n",
    "            # Print sorted `mem_d` values with their contig lengths\n",
    "            print(\"Matching mem_d values sorted by contig length:\")\n",
    "            for mem_d, length in mem_d_lengths:\n",
    "                print(f\"{mem_d}: {length}\")\n",
    "                \n",
    "            # Set new_reference as the longest (last in sorted list)\n",
    "            new_reference = mem_d_lengths[-1][0]  # Get only the `mem_d` part\n",
    "            print(\"new_reference:\", new_reference)\n",
    "            \n",
    "            # Replace the \"rep\" column values where \"rep\" == rep with new_reference\n",
    "            derreplicated_df_fixed.loc[derreplicated_df_fixed[\"rep\"] == rep, \"rep\"] = new_reference\n",
    "\n",
    "derreplicated_df_fixed.to_csv(output_new_clusters, sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "525df0bb",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:53:16.334462Z",
     "iopub.status.busy": "2025-03-18T09:53:16.334217Z",
     "iopub.status.idle": "2025-03-18T09:53:17.553066Z",
     "shell.execute_reply": "2025-03-18T09:53:17.552105Z"
    }
   },
   "outputs": [],
   "source": [
    "# Save representative contig IDs to a CSV file\n",
    "derreplicated_df_fixed.groupby(\"rep\").first().reset_index()[[\"rep\"]].to_csv(output_representatives, index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e8c78af7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-18T09:53:17.558115Z",
     "iopub.status.busy": "2025-03-18T09:53:17.557875Z",
     "iopub.status.idle": "2025-03-18T09:53:18.312840Z",
     "shell.execute_reply": "2025-03-18T09:53:18.312060Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mem_d</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B_C1_V_10_NODE_10_length_2243_cov_6.068099</th>\n",
       "      <td>B_C1_V_10_NODE_10_length_2243_cov_6.068099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_C1_V_10_NODE_47_length_1044_cov_2.564206</th>\n",
       "      <td>B_C1_V_10_NODE_47_length_1044_cov_2.564206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_C1_V_10_NODE_9_length_2279_cov_3.693345</th>\n",
       "      <td>B_C1_V_10_NODE_9_length_2279_cov_3.693345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_C1_V_20_NODE_136_length_1276_cov_3.726454</th>\n",
       "      <td>B_C1_V_20_NODE_136_length_1276_cov_3.726454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_C1_V_20_NODE_138_length_1273_cov_3.242200</th>\n",
       "      <td>B_C1_V_20_NODE_138_length_1273_cov_3.242200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WRB_NODE_7017_length_11561_cov_5.449418</th>\n",
       "      <td>B_OIT4F_V_NODE_18265_length_1149_cov_2.004570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WRB_NODE_7895_length_10645_cov_7.488480</th>\n",
       "      <td>WRB_NODE_7895_length_10645_cov_7.488480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WRB_NODE_8101_length_10455_cov_13.507885</th>\n",
       "      <td>WRB_NODE_8101_length_10455_cov_13.507885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WRB_NODE_817_length_46884_cov_11.202930</th>\n",
       "      <td>RS_CIT4F_V_NODE_5178_length_1301_cov_3.336276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WRB_NODE_8507_length_10107_cov_5.091922</th>\n",
       "      <td>B_OIT2F_V_90_NODE_739_length_2930_cov_4.633043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>229679 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                      mem_d\n",
       "rep                                                                                        \n",
       "B_C1_V_10_NODE_10_length_2243_cov_6.068099       B_C1_V_10_NODE_10_length_2243_cov_6.068099\n",
       "B_C1_V_10_NODE_47_length_1044_cov_2.564206       B_C1_V_10_NODE_47_length_1044_cov_2.564206\n",
       "B_C1_V_10_NODE_9_length_2279_cov_3.693345         B_C1_V_10_NODE_9_length_2279_cov_3.693345\n",
       "B_C1_V_20_NODE_136_length_1276_cov_3.726454     B_C1_V_20_NODE_136_length_1276_cov_3.726454\n",
       "B_C1_V_20_NODE_138_length_1273_cov_3.242200     B_C1_V_20_NODE_138_length_1273_cov_3.242200\n",
       "...                                                                                     ...\n",
       "WRB_NODE_7017_length_11561_cov_5.449418       B_OIT4F_V_NODE_18265_length_1149_cov_2.004570\n",
       "WRB_NODE_7895_length_10645_cov_7.488480             WRB_NODE_7895_length_10645_cov_7.488480\n",
       "WRB_NODE_8101_length_10455_cov_13.507885           WRB_NODE_8101_length_10455_cov_13.507885\n",
       "WRB_NODE_817_length_46884_cov_11.202930       RS_CIT4F_V_NODE_5178_length_1301_cov_3.336276\n",
       "WRB_NODE_8507_length_10107_cov_5.091922      B_OIT2F_V_90_NODE_739_length_2930_cov_4.633043\n",
       "\n",
       "[229679 rows x 1 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "derreplicated_df_fixed.groupby(\"rep\").first()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
